HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,44 @@
+def BFS(sc, graph, name):
+
+    # Note: on the Virtual machine, things work best if everything is partitioned into 2.  
+    
+    # Rdd of name, shortest distance tuples that will be returned.  In each iteration, heros are added to this.  Initialize with the source node, and partition into 2 partitions to match the graph.
+    rdd = sc.parallelize([(name, 0)], 2).cache()
+    
+    # This RDD is used intermediately to build a collection of connections to build on
+    namesRdd = rdd
+
+    # Iterator used in the while loop
+    i = 0
+       
+    acc = sc.accumulator(0)    
+
+    while acc.value == 0:
+
+        i+=1
+
+        # Use join to look up the connections of the next round of connections
+        namesRdd = graph.join(namesRdd)
+
+        # Use mapValues to take just the list of connections and remove the distance
+        namesRdd = namesRdd.mapValues(lambda (x,y): x).partitionBy(2)
+
+        # Take the connections and flatMap them into one big pool to consider
+        namesRdd = namesRdd.values().flatMap(lambda x: x)
+
+        # Reduce the connections to distinct entries
+        namesRdd = namesRdd.distinct()
+        
+        # Subtract out the names that are already in the rdd
+        namesRdd = namesRdd.subtract(rdd.keys())
+        
```

Comment:
Oh, I appreciate I could learn from you those built-in methods in spark. I didn't use either rdd.union nor rdd.substract. If I knew them, I would have tried it. I implemented them in a little bit more complicated way... using set and lists.